src/bson_mapper/bson_archiver.hpp

```diff
@@ -1024,12 +1109,108 @@ inline void CEREAL_SAVE_FUNCTION_NAME(BSONOutputArchive& ar, cereal::NameValuePa
+// the BSONInputArchive will overwrite what was in the destination stdx::optional with a
+// stdx::nullopt, even if there was already a value.
+//
+template <class T,
+          cereal::traits::DisableIf<is_bson<T>::value && !std::is_default_constructible<T>::value> =
+              cereal::traits::sfinae>
+inline void CEREAL_LOAD_FUNCTION_NAME(BSONInputArchive& ar, stdx::optional<T>& t) {
+    if (ar.willSearchYieldValue()) {
+        T value;
+        ar(value);
+        t.emplace(value);
+    } else {
+        t = stdx::nullopt;
+    }
+}
+
+// Overloads for the stdx::optional types wrapping non-default-constructible BSON b_ types.
+// TODO: Perhaps the C++ driver should make all b_ types default constructible.
+template <class T,
+          cereal::traits::EnableIf<is_bson<T>::value && !std::is_default_constructible<T>::value> =
+              cereal::traits::sfinae>
+inline void CEREAL_LOAD_FUNCTION_NAME(BSONInputArchive& ar, stdx::optional<T>& t) {
+    if (ar.willSearchYieldValue()) {
+        typename std::aligned_storage<sizeof(T), alignof(T)>::type value;
+        ar(*reinterpret_cast<T*>(&value));
+        t.emplace(*reinterpret_cast<T*>(&value));
+    } else {
+        t = stdx::nullopt;
+    }
+}
+
+/**
+inline void CEREAL_LOAD_FUNCTION_NAME(BSONInputArchive& ar,
+                                      stdx::optional<bsoncxx::types::b_date>& t) {
+    if (ar.willSearchYieldValue()) {
+        bsoncxx::types::b_date value{std::chrono::system_clock::time_point{}};
+        ar(value);
+        t.emplace(value);
+    } else {
+        t = stdx::nullopt;
+    }
+}
+inline void CEREAL_LOAD_FUNCTION_NAME(BSONInputArchive& ar,
+                                      stdx::optional<bsoncxx::types::b_regex>& t) {
+    if (ar.willSearchYieldValue()) {
+        bsoncxx::types::b_regex value{"", ""};
+        ar(value);
+        t.emplace(value);
+    } else {
+        t = stdx::nullopt;
+    }
+}
+inline void CEREAL_LOAD_FUNCTION_NAME(BSONInputArchive& ar,
+                                      stdx::optional<bsoncxx::types::b_code>& t) {
+    if (ar.willSearchYieldValue()) {
+        bsoncxx::types::b_code value{""};
+        ar(value);
+        t.emplace(value);
+    } else {
+        t = stdx::nullopt;
+    }
+}
+inline void CEREAL_LOAD_FUNCTION_NAME(BSONInputArchive& ar,
+                                      stdx::optional<bsoncxx::types::b_codewscope>& t) {
+    if (ar.willSearchYieldValue()) {
+        bsoncxx::types::b_codewscope value{"", bsoncxx::document::view()};
+        ar(value);
+        t.emplace(value);
+    } else {
+        t = stdx::nullopt;
+    }
+}
+inline void CEREAL_LOAD_FUNCTION_NAME(BSONInputArchive& ar,
+                                      stdx::optional<bsoncxx::types::b_symbol>& t) {
+    if (ar.willSearchYieldValue()) {
+        bsoncxx::types::b_symbol value{""};
+        ar(value);
+        t.emplace(value);
+    } else {
+        t = stdx::nullopt;
+    }
+}*/
```

Comment:
Were these commented-out lines for the non-default-constructible BSON b_ types? 
Can this be removed  since you have the generalized version that uses `std::aligned_storage`?